### DISCUSSION TITLE
Encryption

### DISCUSSION THREAD

**Discussion by peepo5** [stoatchat/discussions#194] [Feature Suggestions]
This is a thread for discussing everything encryption-related

### Edit by maintainer: please use issue https://github.com/revoltchat/revolt/issues/207, thank you!

**Comment by peepo5** [comment]
No mesages lol :<

I think that when encryption is introduced, it must be made mandatory, not optional. I believe this would be better because encryption should be the standard for messaging platforms for privacy and security reasons. This also guarantees that logging cannot happen on the server's end (if implemented correctly).

Adding on to that, metadata should be encrypted (otherwise you will end up in a situation like matrix).
Metadata should be implemented from day 0 otherwise it will be more of a hassle to change the encryption structure.

Encryption should be done for these attributes, including but not limited to:

- Message senders
- Session/device IDs
- Message timestamps
- Room members (join/leave/invite events)
- Message edit events
- Message reactions
- Read receipts
- Nicknames
- Profile pictures
- Room Descriptions

**Reply by insertish** [reply]
Just one thing to point out, mandatory where?
For one, impossible to incredibly difficult for server text channels (redundant if they're public anyways).
For two, what would happen to legacy chats and messages such as in DMs / groups?

**Reply by peepo5** [reply]
Mandatory everywhere.

Legacy chats would (in best case) be backwards compatible so encrypted messages and unencrypted can be resolved in the same chat.

**Reply by BitForger** [reply]
Just spit-balling, but a shared key-pair that is shared with all the clients and associated with the server is an option. I'm sure teams like signal have come up with better solutions.

**Comment by UnixPhonez** [comment]
End to end encryption for private channels would be really nice for web developers who want a secure chat channel room

**Comment by ghost** [comment]
E2EE could be enabled in certain channels just like element does for rooms.

**Reply by peepo5** [reply]
bad idea, always on by default is a better idea for individual 1on1 rooms, due to the optional nature of elements implementation meaning a lot of chats are not encrypted

**Reply by ghost** [reply]
My idea is DMs e2ee by default while groups and server channels are optional.

**Reply by peepo5** [reply]
yea, dm e2e default would be nice, for rooms idk, if there are no performance degredations for large groups, then encryption by default for those would be ideal too. However, it is not just the message content, but metadata which is important

**Reply by ghost** [reply]
Correct, metadata is really sacred.
